src/oneD/StFlow.cpp

```diff
@@ -335,22 +376,119 @@ void StFlow::evalResidual(double* x, double* rsd, int* diag,
             // absorption coefficient for H2O
             if (m_kRadiating[1] != npos) {
                 double k_P_H2O = 0;
-                for (size_t n = 0; n <= 5; n++) {
-                    k_P_H2O += c_H2O[n] * pow(1000 / T(x, j), (double) n);
+                for (int k = 0; k < OPL_table_size; k++) {
+                    if (T(x, j) < TemperatureOPL[k]) {
```

Comment:
This implementation could be improved in terms of computational efficiency as it appears that the same temperature lookup loop is called for each of the species. Beyond, It should be possible to consolidate the species-specific code sections, which - as far as I can tell - are mostly identical except for the species index.

Reply:
PS: I am also not convinced that a linear interpolation within a lookup table is an improvement over a polynomial fit - unless the approximation is poor for the newly added curves.

Reply:
I will make a comparison between old and new curves for H2O and CO2.
Linear interpolation is not an improvement,  but rather a more general assumption. For other species we have not polynomial type assumption and polynomial coefficients. The use of a polynomial can lead to negative values of the absorption coefficients.
I found that CO, CH4 and OH molecules spectra integrated before and presented by me are wrong, so I will change this data.

Reply:
Plots in logarithmic coordinates are at the left part, the same plots in linear coordinates are at the right part.
Results for H2O look similar. Most likely the polynomial approximation was done on a linear scale. So there is a large difference in the logarithmic scale at high temperatures. So linear interpolation in logarithmic scale looks preferable.
The results for CO2 are very different. I will try to find the data from the original source.
![models_check](https://user-images.githubusercontent.com/22547060/106160661-e38c7d80-61a7-11eb-8c0e-5621a03044e6.png)

Reply:
@lavrenyukiv ... thank you - this looks interesting. I was not involved in the original model, so it's hard to tell. I agree that the differences for energy are substantial. The original polynomial coefficients appear to be taken directly from the [TNF link](https://www.sandia.gov/TNF/radiation.html) you had posted (where it states that "_These curve fits were generated for temperatures between 300K and 2500K and may be very inaccurate outside this range._"; the fit itself is for 1000/T so deviations may not look as dramatic?). I am not familiar with RADCAL and assumptions used therein, so some legitimate differences may exist.

*PS:* Incidentally, there was a post on the [User Group](https://groups.google.com/g/cantera-users/c/CCMrMVxTokg/m/BglGYNDEBwAJ) about unphysically low temperatures but it's hard to tell whether these are related as no code was posted.

Reply:
Yes, it is a negative absorption coefficient at low temperatures.
![models_check_3](https://user-images.githubusercontent.com/22547060/106179811-5e13c800-61bd-11eb-8b1d-b0c19ca33a30.png)

Reply:
I am trying to compare spectra.
Results for CO2 from https://github.com/firemodels/radcal/blob/master/Documentation/Figures/CO2_300K.png

![image](https://user-images.githubusercontent.com/22547060/106359766-7bb86d00-6336-11eb-98d2-89581365d5e8.png)

and spectra from hapi
![co2_full](https://user-images.githubusercontent.com/22547060/106359713-41e76680-6336-11eb-8266-84771b9dc823.png)

![co2_crop](https://user-images.githubusercontent.com/22547060/106359725-4875de00-6336-11eb-9b49-d32eb08073a4.png)

![co2_closest](https://user-images.githubusercontent.com/22547060/106359727-4ad83800-6336-11eb-80b5-497c6ddeb87b.png)

Spectra look similar in the specified wavenumber range. The hapi has a greater resolution.